### ISSUE TITLE
Strategy

### ISSUE THREAD

**Issue by jmamma** [yatli/mdhack#1]
@yatli 

I've got the decompiler running and the code in front of me.

What is your strategy / workflow for interpreting the routines. 
How far does the MAME execution proceed?

**Comment by yatli** [comment]
It doesn't get very far, unfortunately..  But it does give you a near sight of what will happen in the future steps, if the registers are properly set up.

I mostly try to interpret the disassembly and C code. Those special addresses (like UART1_BUFFER) are very helpful.

**Comment by yatli** [comment]
You can also try to use "find reference" and see the structure of the code.

**Comment by jmamma** [comment]
do { } while (true);  

at the end of main.

Is everything running on interrupts/timers?

**Comment by yatli** [comment]
Yeah. That'll make things a little more difficult. I'm still learning how to identify the interrupt vectors.

**Comment by yatli** [comment]
The I/O board runs like your ext. screen python script :D

**Comment by jmamma** [comment]
Ha. 7-bit encoding ?

**Comment by yatli** [comment]
Not 7-bit though.
It sends raw coordinates and then followed by framebuffer

**Comment by yatli** [comment]
I'm doing some modification to the memory layout as per MD_MAIN().
```
0x262400 - 0x2B7147 ---- 0-initialized mainOS memory
0x1000088 - 0x1000A29 ---- Copied from 0x26237C -- 0x262D1D (262d1d is the exact end of the unpacked image)
```

**Comment by jmamma** [comment]
How does the UI board know how to distribute the incoming bytes?
Presumably part of the stream will be LCD data, the other part LED update?
Or are LEDs on a separate UART?

**Comment by yatli** [comment]
They're on the same UART and there's a protocol for multiplexing.

**Comment by yatli** [comment]
We're making good progress. After one day searching through the image I've got:

- Timer, I/O board, MIDI ISR
- A lot of GUI drawing routines and data structures, e.g. screen_t, font_t, bitmap_t etc.
- MD menu drawing routines
- MD event handling routines

**Comment by jmamma** [comment]
You are the best @yatli 

Don't stop 😀

**Comment by yatli** [comment]
The whole OS is a gigantic state machine with nested nested nested branching conditions.

One such state variable is accessed by a dozen of routines, interacting with each other.

And the thing works.

They must be secretly using alien compiler or hiring alien programmers.

**Comment by jmamma** [comment]
Isn't that nature of decompiling ? You lose the abstraction that would otherwise be there.

**Comment by yatli** [comment]
You're right. I'm probably looking at well-organized .bss class instances as well.
Let's hope it'll become clear to us later. :)

**Comment by jmamma** [comment]
When do you think you'll have MCL running on the coldfire 😀

**Comment by yatli** [comment]
All I can say is "it's doable now" :)

**Comment by yatli** [comment]
If the flash space is tight, we would need to implement a 2-stage chain bootloader, and load from the midi port

**Comment by yatli** [comment]
this is so addictive.

**Comment by jmamma** [comment]
cyber archaeologist. 

--

I think the possibilities are quite exciting here. Even if the reverse engineer project comes to halt, you've effectively devised a safe way to upload completely new firmwares to the MD. It could run anything we want.

**Comment by yatli** [comment]
@jmamma the bootloader receive upgrade routine is mostly decoded. please take a look if you're interested.

basic format:
handshake + packets

packet:
checksum + target address + payload

two modes: full upgrade/OS image only
full upgrade overwrites everything in flash, including the bootloader

**Comment by jmamma** [comment]
Awesome!

**Comment by jmamma** [comment]
Ill have time to look at this on the weekend.

**Comment by yatli** [comment]
@jmamma The Elektron website is down. Do you have the sysex firmware upgrade around?

**Comment by jmamma** [comment]
didn't come up in my cache.

You should be able to get it from your MD? 

Hold function on boot, select send OS

**Comment by yatli** [comment]
My goal is to emulate the OS upgrade receive process, so that after "receiving" the sysex sequence from the file, it decodes exactly the same .bin file.

Will try to extract the image from MD now.

**Comment by jmamma** [comment]
Yes. I anticipated this would be your next move.
I have to stay 2 moves ahead to keep up 😅

**Comment by yatli** [comment]
;)

Receiving the dump through C6 now; A little bit unsure about the integrity check, could be open-looped.

**Comment by yatli** [comment]
It's probably not the same thing -- 1MB already and still going.

**Comment by yatli** [comment]
[OS_dump_from_MD.zip](https://github.com/yatli/mdhack/files/3934929/OS_dump_from_MD.zip)

Dump complete.

**Comment by yatli** [comment]
Analyze complete.
Decoder uploaded.

From the sysex dump, we decode the exact OS archive as in the .bin dump 0x4000 (elvis included)

**Comment by yatli** [comment]
> cyber archaeologist

Did I put the line somewhere? You're reading my mind.

**Comment by yatli** [comment]
Next move: write the inverse function of UNPACK_FIRMWARE and SYSEX_DECODE.

The latter is very simple.
I'll take a look at firmware packing tomorrow.
Won't be too hard because it's verifiable via the unpack routine.

**Comment by yatli** [comment]
heck, cannot sleep at all.

**Comment by yatli** [comment]
Updated unpack_os a bit with my dialect.

**Comment by yatli** [comment]
The thing looks like a decompression algorithm with two modes:

1. Literal copy
2. Look back copy

**Comment by yatli** [comment]
Now I know this is a variation of the classical deflate algorithm.
See: https://www.zlib.net/feldspar.html

**Comment by yatli** [comment]
compressing the uncompressing code down to <100 lines.
I have something in mind, maybe O(N)

**Comment by yatli** [comment]
Repack algorithm complete.
We will be able to send custom FW very soon!!

@jmamma let's get the m68k compiler toolchain and rock on!

**Comment by yatli** [comment]
I've got compression ratio issues. The original firmware is smaller than what my algorithm can do currently.

Need to do some research on this topic...

**Comment by jmamma** [comment]
That's swedish efficiency at work.

Can you compare the two binaries in a hex editor to see any obvious differences ?

**Comment by yatli** [comment]
Yeah, the difference is pretty obvious. It's how the algorithm chooses to perform look back copy. If too early, it may miss a bigger patch.

For example, say we are at 0x100, and my algorithm greedily chooses to copy 0x100..0x101 from 0x07..0x08

But if we choose to hold back and perform a plain copy at 0x100, we may find a huge patch 0x101..0x200 from 0x0..0x100

**Comment by yatli** [comment]
We have to beat the official image, because the dsp1 image is embarassly right after the os.

**Comment by jmamma** [comment]
"We have to beat the official image, because the dsp1 image is embarassly right after the os."

What do you mean?
